Eliminate an SQL lookup on frontend cached element partials

mtomov opened the pull request:
* when not using page caching

   When page_cache is not turned off, it is unnecessary to store the
   element within the page it is being rendered using the `store_page`
   method. Thereby, the method call is unnecessary, and avoiding it
   saves an SQL query on every cached element render.

```
     Alchemy::Page Exists (2.2ms)  SELECT  DISTINCT 1 AS one FROM `alchemy_pages` INNER JOIN `alchemy_elements_alchemy_pages` ON `alchemy_pages`.`id` = `alchemy_elements_alchemy_pages`.`page_id` WHERE `alchemy_elements_alchemy_pages`.`element_id` = 569 AND `alchemy_pages`.`id` = 2 LIMIT 1
```

  So, if a non-cached page with N elements would result in N such queries.

  This change saves those Exists lookups.

Let me know if you prefer to place the pull request against 3.2 instead of master. 

Solves AlchemyCMS/alchemy_cms/issues/898
Thank you!

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://mtomov-eliminate-element-db-lookup-query.alchemycms-alchemy-cms-42e8f4fc945384e0f6b5.ttrcloud.com/admin' target='_blank'>Open stage</a> or <a href='http://mtomov-eliminate-element-db-lookup-query.alchemycms-alchemy-cms-42e8f4fc945384e0f6b5.ttrcloud.com/admin/teatro.log' target='_blank'>view logs</a>.

tvdeyen commented on app/helpers/alchemy/elements_helper.rb:
```diff
@@ -159,7 +159,7 @@ def render_element(element, part = :view, options = {}, counter = 1)
       }
 
-      element.store_page(@page) if part.to_sym == :view
+      element.store_page(@page) if part.to_sym == :view && Alchemy::Config.get(:cache_pages)
```
We should ask if the `@page` needs to be cached or not. Using the global setting is ok, but will very unlikely turned off for the most sites.

mtomov commented:
Yes, valid point.

I have moved `cache_page?` method onto Page itself. Let me know what you think. I've removed only one test from the original ones, which I think can't happen in the controller due to the `authorize .. @page` at the top of the `show` method.

```ruby
      it 'returns false when there is no page' do
        controller.instance_variable_set('@page', nil)
        expect(subject).to be false
      end
```

Let me know if you prefer to name the method differently (just `cache?`), or maybe to split the commits into one, which moves the method onto Page, and the other, which fixes this issue, or maybe squash them both.

tvdeyen commented on app/helpers/alchemy/elements_helper.rb:
```diff
@@ -159,7 +159,7 @@ def render_element(element, part = :view, options = {}, counter = 1)
       }
 
-      element.store_page(@page) if part.to_sym == :view
+      element.store_page(@page) if part.to_sym == :view && @page.cache_page?
```
I think now that the `if` condition is having 2 conditionals, we shouldn't use a postfix `if` here anymore.

It's harder to scan the rather long method.

tvdeyen commented on app/models/alchemy/page.rb:
```diff
@@ -361,6 +361,34 @@ def update_node!(node)
       update_columns(hash)
     end
 
+
```
Yap! Could you please remove it? Thanks

tvdeyen commented on app/models/alchemy/page.rb:
```diff
@@ -361,6 +361,34 @@ def update_node!(node)
+    #       cache: false
+    #
+    # Note: This only sets the cache control headers and skips rendering of the page body, if the cache is fresh.
+    # This does not disable the fragment caching in the views. So if you don't want a page and it's elements to be cached,
```
Would you mind to add proper line breaks to the comments. Not you're fault, I know, but as you are at here anyway?

tvdeyen commented on app/models/alchemy/page.rb:
```diff
@@ -361,6 +361,34 @@ def update_node!(node)
+    #
+    def cache_page?
+      return false if !Alchemy::Config.get(:cache_pages) ||
+        !Rails.application.config.action_controller.perform_caching
```
I would prefer a new private method `caching_enabled?` and an `unless` for this condition.

tvdeyen commented on spec/models/page_spec.rb:
```diff
@@ -1737,6 +1737,43 @@ module Alchemy
+      subject { page.cache_page? }
+
+      before do
+        Rails.application.config.action_controller.perform_caching = true
```
Please don't forget to reset it to the former value after this test.

tvdeyen commented:
Great! Thanks

I think the method name `cache_page?`is ok, despite the duplication. But I would like to see the method in the `page_natures` module. There are already lots of methods concerning the character/nature of a page.

tvdeyen commented:
@robinboening @mamhoff have any position on this?

tvdeyen commented:
Ah, and yes, it would be great if the two commits have reversed order. First the method extraction, then the sql query guard.

Thanks again for your contribution :cake:

mtomov commented:
Thanks for the feedback. I've been a bit busy those days, but will try to finish both this one and the other pull requests. Thank you!

mtomov commented:
Address all points! Let me know if anything else.

mtomov commented:
You sure?, because github says: "This branch has no conflicts with the base branch", and I think it's rebased.

tvdeyen commented:
Hmmm this strange. Maybe my iOS client had outdated data. Works

Thanks again for your contribution 🚀

mtomov commented:
Great! I'm happy to help improve!

On Sun, Jan 24, 2016 at 3:05 PM, Thomas von Deyen <notifications@github.com>
wrote:

> Hmmm this strange. Maybe my iOS client had outdated data. Works
>
> Thanks again for your contribution 🚀
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/AlchemyCMS/alchemy_cms/pull/911#issuecomment-174307106>
> .
>

mtomov commented:
Hm, actually this won't work. 

`touchable_pages` is the only way an update to an element also changes the updated_at time of the parent page as per [touching.rb](https://github.com/AlchemyCMS/alchemy_cms/blob/master/lib/alchemy/touching.rb#L32). So, if `#store_page` is disabled, then the updated_at time will always be stale, and [Page#cache_key with current_preview == self](https://github.com/AlchemyCMS/alchemy_cms/blob/master/app/models/alchemy/page/page_natures.rb#L97) will never render .. so the admin preview is broken.

So, there's a few ways to address that. One is to always update the element's parent (`touch: true` or in the `touch_pages`'s method). Another is to revert all of that.. What's your thoughts?

Btw, why is the Publish button disabled for Global Pages? I cache them like any other page with `cache (global_page) do `, but without the button, admin users can't invalidate them. Am I missing something?

tvdeyen commented:
Currently I don't have to time to look into this, so I will revert edffb37fa1530dce1736b7ee600a988cd7616c92 until we have a solution for this.

mtomov commented:
:+1: thanks. sorry about that.

tvdeyen commented:
All, fine. No worries.
